File: widgets/_lab/mutational-signatures-reference.R
# Widget 70's engine reference: R's NMF 0.28 runs `brunet` — the method
# `extractSignatures` runs — from a fixed start for exactly 200 iterations on a
# synthetic count matrix plus pConstant 0.1, and the result is written beside
# the input so `_lab/mutational-signatures-verify.mjs` § 1 can hold widget
# 41's `updateKL`, with brunet's floor, to it.
#
#   node widgets/_lab/mutational-signatures-reference.mjs <scratch>/ms-ref-in.tsv
#   Rscript widgets/_lab/mutational-signatures-reference.R <scratch>/ms-ref-in.tsv widgets/_lab/mutational-signatures-reference.tsv
#
# The input is the widget's own simulated cohort (nothing TCGA's). NMF was
# installed from CRAN on Kenneth's approval, 2026-09-18. A start is passed as
# an NMF model (`seed = nmfModel(W0, H0)`), and `nmf.stop.iteration(200)`
# replaces brunet's own stopping rule, so both sides run the same count.
suppressPackageStartupMessages(library(NMF))
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 2) stop("usage: Rscript mutational-signatures-reference.R <in.tsv> <out.tsv>")
d <- read.delim(args[1], colClasses = c("character", "integer", "integer", "numeric"))
block <- function(name) {
  b <- d[d$block == name, ]
  m <- matrix(0, max(b$i), max(b$j))
  m[cbind(b$i, b$j)] <- b$value
  m
}
V <- block("V"); W0 <- block("W0"); H0 <- block("H0")
ITER <- 200
fit <- nmf(V + 0.1, ncol(W0), method = "brunet", seed = nmfModel(W = W0, H = H0), .stop = nmf.stop.iteration(ITER))
stopifnot(niter(fit) == ITER)
W <- basis(fit); H <- coef(fit)
long <- function(name, m) data.frame(block = name, i = rep(seq_len(nrow(m)), ncol(m)), j = rep(seq_len(ncol(m)), each = nrow(m)),
                                     value = sprintf("%.17g", as.vector(m)))
out <- rbind(long("V", V), long("W0", W0), long("H0", H0), long("W", W), long("H", H))
con <- file(args[2], open = "wb")
writeLines(c(sprintf("# NMF %s, method brunet, %d iterations from W0/H0 on V + 0.1; %s", packageVersion("NMF"), ITER, R.version.string),
             "block\ti\tj\tvalue", paste(out$block, out$i, out$j, out$value, sep = "\t")), con, sep = "\n")
close(con)
cat(sprintf("brunet: %d iterations, KL residual %.6f; wrote %d values\n", niter(fit), residuals(fit), nrow(out)))

File: widgets/_lab/mutational-signatures-reference.mjs
/* The input for widget 70's engine reference: a synthetic 96 × 24 count
 * matrix — the first 24 tumors of the widget's own cohort at seed 1, nothing
 * TCGA's — and a starting W (96 × 4) and H (4 × 24) drawn from a fixed seed.
 *
 *   node widgets/_lab/mutational-signatures-reference.mjs <scratch>/ms-ref-in.tsv
 *   Rscript widgets/_lab/mutational-signatures-reference.R <scratch>/ms-ref-in.tsv widgets/_lab/mutational-signatures-reference.tsv
 *
 * R's NMF 0.28 then runs `brunet` from that start for exactly 200 iterations
 * on the counts plus pConstant 0.1, and writes the input and the result into
 * one file, which `_lab/mutational-signatures-verify.mjs` § 1 reads.
 * Long format: block, i, j, value, with 17 significant digits.
 */
import { writeFileSync } from "node:fs";
import { cohortFor } from "../mutational-signatures/model.js";
import { makeRng } from "../core/rng.js";

const out = process.argv[2];
if (!out) { console.error("usage: node mutational-signatures-reference.mjs <out.tsv>"); process.exit(1); }
export const REF_TUMORS = 24, REF_RANK = 4, REF_SEED = 20260919;

const co = cohortFor(1);
const V = co.M.map((row) => Array.from(row).slice(0, REF_TUMORS));
const rng = makeRng(REF_SEED);
const W0 = Array.from({ length: 96 }, () => Array.from({ length: REF_RANK }, () => rng.next() + 0.01));
const H0 = Array.from({ length: REF_RANK }, () => Array.from({ length: REF_TUMORS }, () => rng.next() + 0.01));
const lines = ["block\ti\tj\tvalue"];
const put = (name, A) => A.forEach((row, i) => row.forEach((v, j) => lines.push(`${name}\t${i + 1}\t${j + 1}\t${Number(v).toPrecision(17)}`)));
put("V", V);
put("W0", W0);
put("H0", H0);
writeFileSync(out, `${lines.join("\n")}\n`);
console.log(`wrote ${lines.length - 1} values to ${out}`);
